src/main/java/org/variantsync/diffdetective/variationtree/VariationTreeNode.java

```diff
@@ -0,0 +1,364 @@
+     * mapping is {@code "true"}). The label of this node is empty and therefore the line numbers
+     * are {@link DiffLineNumber#InvalidLineNumber invalid}.
+     *
+     * @see addChild
+     */
+    public static VariationTreeNode createRoot() {
+        return new VariationTreeNode(
+                NodeType.IF,
+                DiffLineNumber.InvalidLineNumber,
+                DiffLineNumber.InvalidLineNumber,
+                FixTrueFalse.True,
+                new ArrayList<>()
+        );
+    }
+
+    /**
+     * Convenience constructor for creating an artifact node of an {@link VariationTree}.
+     *
+     * @param fromLine the start line number of the code of {@code label}
+     * @param toLine the end line number (exclusive) of the code of {@code label}
+     * @param label a list of lines used as label
+     * @see addBelow
+     */
+    public static VariationTreeNode createArtifact(int fromLine, int toLine, List<String> label) {
+        return new VariationTreeNode(NodeType.ARTIFACT, fromLine, toLine, null, label);
+    }
+
+    @Override
+    public VariationTreeNode getBackingNode() {
+        return this;
+    }
+
+    @Override
+    public NodeType getNodeType() {
+        return nodeType;
+    }
+
+    @Override
+    public List<String> getLabelLines() {
+        return label;
+    }
+
+    /**
+     * Replaces the current label by {@code newLabelLines}.
+     *
+     * <p>The given list is not copied, so modifications of {@code newLabelLines} will be visible by
+     * {@link getLabelLines}.
+     *
+     * @see getLabelLines
+     */
+    public void setLabelLines(List<String> newLabelLines) {
+        label = newLabelLines;
+    }
+
+    /**
+     * Adds the given lines to the source code lines of this node.
+     *
+     * @param lines lines to add
+     * @see getLabelLines
+     */
+    public void addLines(final List<String> lines) {
+        this.label.addAll(lines);
+    }
+
+    @Override
+    public int getFromLine() {
+        return from;
+    }
+
+    @Override
+    public void setFromLine(int from) {
+        this.from = from;
+    }
+
+    @Override
+    public int getToLine() {
+        return to;
+    }
+
+    @Override
+    public void setToLine(int to) {
+        this.to = to;
+    }
+
+    @Override
+    public VariationNode<VariationTreeNode> getParent() {
+        return parent;
+    }
+
+    /**
+     * Sets {@code newParent} as the new parent of this node.
+     *
+     * <p>This node must not have a parent yet.
+     */
+    private void setParent(final VariationTreeNode newParent) {
+        Assert.assertTrue(parent == null);
+        this.parent = newParent;
+    }
+
+    @Override
+    public List<VariationNode<VariationTreeNode>> getChildren() {
+        return Collections.unmodifiableList(childOrder);
+    }
+
+    @Override
+    public void addChild(final VariationNode<VariationTreeNode> child) {
```

Comment:
So to all these methods, we could also pass other nodes that wrap `VariationTreeNode`s. Is this something we want to allow? What about making these methods only accept a value of the backing type (here: VariationTreeNode)? Would this break something with the Projections or would it be a fine restriction?

Reply:
It would also avoid to invoke `getBackingNode` everywhere which could sometimes easily be forgotten (see below).

Reply:
This would shift the burden of calling `getBackingNode` from the implementation to the caller. That makes the interface considerably less simple.
I currently can't think of new problems that can be cause by mixing such views. I say "new" because all the problems I can imagine are also possible by just mixing two unconnected trees of `DiffNode`s.

After thinking about our discussion about the check in the constructor today, I got another idea:
There is a design pattern in C++ called [curiously recurring template pattern (CRTP)](https://en.wikipedia.org/wiki/Curiously_recurring_template_pattern) where a base class (in our case `VariationNode`) receives the derived class itself (in our case `VariationTreeNode` and `Projection`) as template parameter.
We could use the same design pattern and even enforce it at compile time using the following:
```java
public abstract class VariationNode<T extends VariationNode<T>> implements HasNodeType {
```

This would give us the following benefits:
- We can use the type parameter `T` instead of `VariationNode<T>` everywhere,
- ensuring that no two views are compatible,
- nicer function signatures (though, we should use a different type variable name then) and
- way less invocations of `getBackingNode`.

There are also the following drawbacks:
- Generic functions outside of `VariationNode` have to add a constraint to their type variable:
  before: `public <T> fancyAlgorithm(VariationNode<T> node)`
  after: `public <T extends VariationNode<T>> fancyAlgorithm(T node)`
  This may be confusing for new users. However, I think this is a minor problem as this can be documented and the compiler is able to give a pretty useful error message (giving the correct constraint).

The implementation of this is added as another commit so you can review is separately.

Reply:
See this commit: 1cb04cecebcbcb7426d576f8716c0dafa6e31ef9

Reply:
I actually think the code is now easier to read because logic for data and view is not mixed anymore. In particular, we do not have `Projection extends VariationNode<DiffNode>` anymore which involves three types.